### PR TITLE
Route Alerts

### PR DESCRIPTION
**Pull request by Udumft**
Adds Route Alerts processing and deploying it to user.

Implementation is based on entities, provided by `MapboxDirections` and reported by `Navigation Native`.
`Incidents` support is subject to be changed.

This PR depends on [Updating Navigation Native to v22](https://github.com/mapbox/mapbox-navigation-ios/pull/2648) as it needs to further update it to v24. Also, MapboxDirections should be [updated](https://github.com/mapbox/mapbox-directions-swift/pull/466) to reflect `Route Alerts` data.

I had to update Routing Tiles fixture because the old one was apparently quite old and did not contain `admin. iso_3166_1_alpha3` name which became obligatory. Updated tiles take much more disk space so I wonder if this can be reduced somehow...

### CODE REVIEW

**Comment by avi-c** [comment]
I have successfully updated Apex to use this branch and the MapboxDirections branch it pulls in. I then used it to find and label all the TollBooth features.

File: MapboxCoreNavigation/RouteProgress.swift
Comment (avi-c) [inline MapboxCoreNavigation/RouteProgress.swift:152]: Are we able to provide any information on how far ahead "upcoming" is supposed to be? Is it tunable or related to speed of travel or current step?

File: Cartfile
Comment (avi-c) [inline Cartfile:5]: Will we be moving to a specific version once the directions changes are merged into MapboxDirections?

File: Cartfile
Comment (avi-c) [inline Cartfile:2]: Maybe we want v25.0 now?

File: MapboxCoreNavigation/RouteController.swift
Comment (avi-c) [inline MapboxCoreNavigation/RouteController.swift:339]: Are alerts always present or do we want this to be an optional?

File: Cartfile
Comment (Udumft) [reply Cartfile:2]: Actually, yes, you are right. In v25 NN [renamed Incident type](https://github.com/mapbox/mapbox-navigation-native/pull/1787) JSON key which affects the feature.

File: Cartfile
Comment (Udumft) [reply Cartfile:5]: Correct. This is temporary while Directions didn't release with [corresponding update](https://github.com/mapbox/mapbox-directions-swift/pull/466).

File: MapboxCoreNavigation/RouteController.swift
Comment (Udumft) [reply MapboxCoreNavigation/RouteController.swift:339]: Array itself is non-nil. It is also important to update it each time since "upcoming" alerts do not include already reported ones, so most of the time this array is reset to empty.

File: MapboxCoreNavigation/RouteProgress.swift
Comment (Udumft) [reply MapboxCoreNavigation/RouteProgress.swift:152]: Good question. At least I am not aware of such configurations, and this is decided by NN. I'll investigate

Comment (Udumft) [reply MapboxCoreNavigation/RouteProgress.swift:152]: So I found the doc should be a bit corrected: `upcomingRouteAlerts` contain all events, that user still didn't pass entirely. In other words, at the beginning of the route this array holds all events from all the route. Event is removed from this array only when user has passed it's `end coordinate`

**Comment by avi-c** [comment]
@MaximAlien @1ec5 if possible it would be nice to get this merged today so we can be closer to Android parity

**Comment by MaximAlien** [comment]
> @MaximAlien @1ec5 if possible it would be nice to get this merged today so we can be closer to Android parity

@avi-c, not sure we can do this just yet. Since this PR is based on Navigation Native update (which currently fails when running CocoaPods linter on Xcode 12) I think it'd make sense to wait for a fix in Mapbox Common and Navigation Native. I've created ticket here: https://github.com/mapbox/mapbox-sdk-common/issues/570. As I also can see this PR updates to Navigation Native 24.0.0, it'll have to contain same fix as well.

I can also see that there are several compilation errors like:
```
/Users/distiller/project/MapboxCoreNavigation/BorderCrossingInfo.swift:6:11: cannot find type 'AdministrationRegion' in scope
```

**Comment by Udumft** [comment]
About compile error mentioned above: this is related to `mapbox-directions` not yet released [required changes](https://github.com/mapbox/mapbox-directions-swift/pull/466), so cocoapods is unable to checkout correct source code.

**Comment by avi-c** [comment]
@MaximAlien As @Udumft points out above, getting a review on https://github.com/mapbox/mapbox-directions-swift/pull/466 and getting it merged would be great to unblock this one.

**Comment by avi-c** [comment]
I just merged https://github.com/mapbox/mapbox-directions-swift/pull/466 so the Intersection attributes are present now.

**Comment by 1ec5** [comment]
> I had to update Routing Tiles fixture because the old one was apparently quite old and did not contain `admin. iso_3166_1_alpha3` name which became obligatory. Updated tiles take much more disk space so I wonder if this can be reduced somehow...

It turns out this is the root cause of mapbox/mapbox-directions-swift#491. mapbox/mapbox-directions-swift#492 works around the issue by making the alpha-3 property optional. Per mapbox/mapbox-directions-swift#477, Swift code on Apple platfoms is unlikely to need the alpha-3 code anyways.

**Review by MaximAlien — COMMENTED** [review]
It's been a while since this PR was updated. I think good idea would be to rebase onto `main` to get latest changes.

File: MapboxCoreNavigationTests/CocoaPodsTest/PodInstall/Podfile.lock
Comment (Udumft) [inline MapboxCoreNavigationTests/CocoaPodsTest/PodInstall/Podfile.lock:23]: Oddly, NN v25 uses older `MapboxCommon` version than v22. Could this be a reason for CI Pod check? Verbose output shows an issue with linking `MapboxCommon` slices:
```
-> MapboxCoreNavigation (1.2.0-alpha.1)
    - ERROR | [iOS] xcodebuild: Returned an unsuccessful exit code.
    - NOTE  | xcodebuild:  note: Using new build system
    - NOTE  | xcodebuild:  note: Building targets in parallel
    - NOTE  | xcodebuild:  note: Using codesigning identity override: -
    - NOTE  | [iOS] xcodebuild:  note: Planning build
    - NOTE  | [iOS] xcodebuild:  note: Constructing build description
    - NOTE  | [iOS] xcodebuild:  warning: Skipping code signing because the target does not have an Info.plist file and one is not being generated automatically. (in target 'App' from project 'App')
    - NOTE  | [iOS] xcodebuild:  warning: [CP] Unable to find matching .xcframework slice in 'MapboxCommon/MapboxCommon.xcframework MapboxCommon framework ios-i386_x86_64-simulator ios-armv7_arm64 ios-x86_64-maccatalyst' for the current build architectures (arm64 x86_64 i386).
    - NOTE  | xcodebuild:  ld: warning: directory not found for option '-F/Users/viktorkononov/Library/Developer/Xcode/DerivedData/App-dxvmejvrbgstrehborjnnbpvaujh/Build/Products/Release-iphonesimulator/XCFrameworkIntermediates/MapboxCommon'
    - NOTE  | xcodebuild:  clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by Udumft** [comment]
I had to update Nav Native to version 26.3 due to some dependencies mismatch: NN v25 (where RouteAlerts feature originates) used `MapboxCommons` v7.1.1 which causes pods issues. According to the schedule, we were to update it anyways for NavSDK v1.2

File: MapboxCoreNavigation/BorderCrossingInfo.swift
Comment (1ec5) [inline MapboxCoreNavigation/BorderCrossingInfo.swift:17]: Is it expected that there will be more information about the border crossing, or could it be more efficiently represented as a tuple?

File: MapboxCoreNavigation/RouteProgress.swift
Comment (1ec5) [inline MapboxCoreNavigation/RouteProgress.swift:150]:
```suggestion
     The contents of the array depend on user's current progress along the route and are modified on each location update. This array contains only the alerts that the user has not passed. Some events may have non-zero length and are also included while the user is traversing it. You can use this property to get information about incoming points of interest.
```

File: MapboxCoreNavigation/TunnelInfo.swift
Comment (1ec5) [inline MapboxCoreNavigation/TunnelInfo.swift:9]: Since this struct stores only a name, do we need the struct, or can we inline the name?

If we do expect to extend this type with more information in the future, then consider calling it Tunnel. (Compare to MapboxDirections’ [Road](https://github.com/mapbox/mapbox-directions-swift/blob/4184df3c07694f75738a99728cdfa5504c8e4632/Sources/MapboxDirections/RouteStep.swift#L307) struct.)

File: MapboxCoreNavigation/UpcomingRouteAlertInfo.swift
Comment (1ec5) [inline MapboxCoreNavigation/UpcomingRouteAlertInfo.swift:16]: This enumeration indicates more than just the type, so we should rename it to just `RouteAlert`. There will probably be a conflict with MapboxNavigationNative, but we can solve that conflict by selective imports and namespaces. In the worst-case scenario, a separate file could import MapboxNavigationNative and declare `InternalRouteAlert` as a type alias of `MapboxNavigationNative.RouteAlert`.

File: MapboxCoreNavigation/UpcomingRouteAlertInfo.swift
Comment (1ec5) [inline MapboxCoreNavigation/UpcomingRouteAlertInfo.swift:9]: Is “upcoming” the right terminology to use? Is there a situation where the developer might end up referring to a past UpcomingRouteAlert? This is sort of like a “positioned” route alert in that it combines the alert itself with positioning information.

File: MapboxCoreNavigation/UpcomingRouteAlertInfo.swift
Comment (1ec5) [inline MapboxCoreNavigation/UpcomingRouteAlertInfo.swift:63]: Is this initializer failable because the alert may come back with no metadata? Or are these guard statements meant to guard against internal inconsistency? If the latter, perhaps these code paths should be `preconditionFailure`s.

File: MapboxCoreNavigation/IncidentInfo.swift
Comment (1ec5) [inline MapboxCoreNavigation/IncidentInfo.swift:20]: mapbox/mapbox-directions-swift#482 will simplify this initializer a bit.

**Review by 1ec5 — COMMENTED** [review]
Also remember to update the changelog to mention the added types and properties, as well as the upgraded MapboxDirections and MapboxNavigationNative dependencies.

File: MapboxCoreNavigation/BorderCrossingInfo.swift
Comment (Udumft) [reply MapboxCoreNavigation/BorderCrossingInfo.swift:17]: To be honest: I don't know. At least now having it as `struct` is justified by introducing a convenience initializer which makes it easier to use than a `tuple`.

File: MapboxCoreNavigation/TunnelInfo.swift
Comment (Udumft) [reply MapboxCoreNavigation/TunnelInfo.swift:9]: Situation here is same as for `BorderCrossingInfo`. I wanted to have similar configuration for each Route Alert possible. IMHO if we have some mix of `struct`s, `tuple`s and inlined values that would look inconsistent. That's why I decided to stick with `struct`s everywhere because this is the most universal approach and it allows extending functionality and data without breaking backwards compatibility.

File: MapboxCoreNavigation/UpcomingRouteAlertInfo.swift
Comment (Udumft) [reply MapboxCoreNavigation/UpcomingRouteAlertInfo.swift:9]: `UpcomingRouteAlertInfo` are provided to a user on location updates and the feed contains only not passed events. So for the context of how SDK supplies and uses it, "upcoming" seems correct.

File: MapboxCoreNavigation/UpcomingRouteAlertInfo.swift
Comment (Udumft) [reply MapboxCoreNavigation/UpcomingRouteAlertInfo.swift:63]: Hm... Originally I decided to make this one failable to cover cases when NN submits unknown alert type, so that SDK could just omit it. But it seems that such cases are impossible because NN version will always be in sync with SDK version, so having `default` case here might even become a potential point to forget add support for new alert types.
About optional alert info structures - these should not be nil if alert type is correct, so I think yes, that might be considered as an error.

File: MapboxCoreNavigation/UpcomingRouteAlertInfo.swift
Comment (1ec5) [reply MapboxCoreNavigation/UpcomingRouteAlertInfo.swift:9]: I think it’s important for the name of a public type to be intrinsic to the data contained inside of it; to some extent, the name should be independent of the situations in which the type is or isn’t used. It would be awkward if the application keeps its own array of passed “upcoming” alerts. (For a contrived example, consider a counter of how many “upcoming” alerts the user has already passed.)

File: MapboxCoreNavigation/UpcomingRouteAlertInfo.swift
Comment (1ec5) [reply MapboxCoreNavigation/UpcomingRouteAlertInfo.swift]: Ideally we’d add default unknown cases, but unfortunately, MapboxNavigationNative incorrectly declares all its structs as frozen.

Comment (Udumft) [reply MapboxCoreNavigation/UpcomingRouteAlertInfo.swift]: I thought that having a `default` case might create a room for an error once any new alert type is added to NN. If such happens now - compiler will alarm.

File: MapboxCoreNavigation/UpcomingRouteAlertInfo.swift
Comment (1ec5) [inline MapboxCoreNavigation/UpcomingRouteAlertInfo.swift:55]: Is it possible to add a default or default unknown case even though the compiler thinks that the enumeration is frozen and the switch statement is exhaustive?

File: MapboxCoreNavigation/UpcomingRouteAlertInfo.swift
Comment (1ec5) [reply MapboxCoreNavigation/UpcomingRouteAlertInfo.swift]: Will it reliably produce an error if MapboxNavigationNative enables module stability?

File: MapboxCoreNavigation/UpcomingRouteAlertInfo.swift
Comment (Udumft) [reply MapboxCoreNavigation/UpcomingRouteAlertInfo.swift:55]: Adding `@unknown default` case triggers a warning since all cases are already handled and `RouteTypeAlert` is frozen ATM. If any issues arise with module stability and bindgen products - it will affect not just this one switch, and more general approach would be required.

File: CHANGELOG.md
Comment (Udumft) [reply CHANGELOG.md]: Done

File: CHANGELOG.md
Comment (1ec5) [inline CHANGELOG.md:7]:
```suggestion
* Increased the minimum versions of `MapboxNavigationNative` to v26.3, `MapboxCommon` to v9.1.0 and `MapboxDirections` to v1.2. ([#2694](https://github.com/mapbox/mapbox-navigation-ios/pull/2694))
```

File: MapboxCoreNavigation/RouteAlert.swift
Comment (1ec5) [inline MapboxCoreNavigation/RouteAlert.swift:17]: This enumeration and its cases should be documented, since the public `alert` property uses them. It can be just a perfunctory documentation comment, such as “Incident with the given incident details” for `incident`.

**Review by 1ec5 — APPROVED** [review]
I’ve pushed MapboxDirections v1.2.0-alpha.3 to CocoaPods trunk, so the CocoaPods integration tests should pass after rerunning. Thanks for your persistence on this feature!

File: CHANGELOG.md
Comment (Udumft) [reply CHANGELOG.md]: Changed

File: MapboxCoreNavigation/RouteAlert.swift
Comment (Udumft) [reply MapboxCoreNavigation/RouteAlert.swift]: Doc comments added

**Pull request merged**
